### PR TITLE
DDF-2824 Solr documentation and service update

### PR DESCRIPTION
**Pull request by ahoffer**
#### What does this PR do?
The purpose of this PR is to make it easier to run DDF and Solr as services managed by the operating system. There are changes in 3 areas:
* The Windows file ddf.bat is decomposed into two files, ddf.bat, and ddfsolr.bat. The later file only manages starting, stopping, and restarting Solr
* A new file, solr.service, is provided for operating systems using the systemd init manager
* The documentation about managed services has been restructured and completely rewritten.

#### Who is reviewing it? 
@ryeats @mcalcote @oconnormi 

#### Ask 2 committers to review/merge the PR and tag them here.
@clockard
@figliold
@ricklarsen - Documentation
@rzwiefel

#### How should this be tested?
This PR should be heroed on 
* Windows
* CentOS 7
* and maybe a recent Ubuntu distro

To test the PR, open the documentation to section 8.2. Managed Services, and follow the instructions.

#### Checklist:
- [X] Documentation Updated
- [NA] Update / Add Unit Tests
- [NA] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

**Comment by ahoffer** [comment]
build now

File: distribution/docs/src/main/resources/content/_installing/os-services.adoc [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+:summary: Install Processes as Services.
+:project: ${branding}
+:order: 03
+
+=== {title}
+
+The lifecycle of ${branding} and Solr processes can be managed by the operating
+system. The procedures for installing a processes as a service varies by operating
+system and the particular init manager used by the operating system.
+The ${branding} documentation provides instructions to install
```
Comment (figliold) [inline distribution/docs/src/main/resources/content/_installing/os-services.adoc]: ❓ Should that say "The ${branding} documentation provides instructions to install ${branding} as a managed service for supported platforms."?

File: distribution/docs/src/main/resources/content/_installing/os-services.adoc [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+${branding} and as managed services for supported platform. However, the
+documentation cannot account for all possible environments. Please consult the
```
Comment (figliold) [inline distribution/docs/src/main/resources/content/_installing/os-services.adoc]: :question: Should this say "... and its init manager if the instructions in this document are inadequate."?

File: distribution/docs/src/main/resources/content/_installing/solr-service.adoc [hunk 3]
```diff
@@ -0,0 +1,72 @@
+:title: Run Solr as Managed Service
+:type: startingIntro
+:status: published
+:summary: Install Solr as a Service.
+:project: ${branding}
+:order: 05
+
+==== {title}
+
+These instructions are configuring Solr as a service managed by the operating system. A common
```
Comment (figliold) [inline distribution/docs/src/main/resources/content/_installing/solr-service.adoc]: ❓ Should this say "These instructions are for configuring ..."?

File: distribution/docs/src/main/resources/content/_installing/solr-service.adoc [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+reason is to start Solr when the system boots.
+
+===== Configure Solr as a Systemd Service
+These instructions are for the unix family of operating systems running the systemd
```
Comment (figliold) [inline distribution/docs/src/main/resources/content/_installing/solr-service.adoc]: ✏️ Unix?

File: distribution/docs/src/main/resources/content/_installing/solr-service.adoc [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+ init manager. If configuring a Windows system, see
+ <<{managing-prefix}configure_solr_as_a_windows_service,Configure Solr as a Windows Service>>
+
+
+. From the operating system command line, enable a Solr service using a provided configuration file.
+Use the full path to the file.
++
+----
+ systemctl enable <${branding}_HOME>/solr/service/solr.service>
```
Comment (emmberk) [inline distribution/docs/src/main/resources/content/_installing/solr-service.adoc]: ✏️ extra `>` at the end of this line

File: distribution/docs/src/main/resources/content/_installing/os-services.adoc [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+documentation for the operating system and its init manager for if the instructions
+in this document are inadequate.
+
+==== Procedure for using managed services
+. If ${branding} is running, stop it.
+. Edit `<${branding}_HOME>/etc/system.properties` and set `start.solr=false`.
+This prevents the ${branding} scripts from attempting to manage Solr's lifecycle.
+. <<{managing-prefix}run_solr_as_managed_service,Configure Solr to run as a managed service>>.
+. Start ${branding} and verify it is communicating correctly with Solr. To verify, first log out
+of ${branding}. Then navigate to ${catalog_ui} (login when prompted). If ${catalog_ui} opens
```
Comment (emmberk) [inline distribution/docs/src/main/resources/content/_installing/os-services.adoc]: These properties didn't get replaced correctly
image

File: distribution/docs/src/main/resources/content/_installing/solr-service.adoc (hunk 4 above)
Comment (ahoffer) [reply distribution/docs/src/main/resources/content/_installing/solr-service.adoc]: I had to google this one.  I found different advice, but this made the most sense to me:
http://www.greens.org/about/unix.html
@ricklarsen What would you like to see?

File: distribution/docs/src/main/resources/content/_sources/solr-catalog-provider.adoc [hunk 7]
```diff
@@ -75,12 +68,6 @@ Zookeeper helps manage the overall structure.
  +------------------+   +---------------------+
 ....
 
-[NOTE]
-====
-Solr Cloud is currently in Beta version.
-Bugs and feature changes may affect the final release.
-====
-
```
Comment (ahoffer) [inline distribution/docs/src/main/resources/content/_sources/solr-catalog-provider.adoc]: It was not a good idea to document the settings both here and in global-setting.adoc. Removed.

File: distribution/docs/src/main/resources/content/_sources/solr-catalog-provider.adoc [hunk 8]
```diff
@@ -32,13 +32,6 @@ No installation is required because ${branding} includes a distribution of Apach
 .Configuring Solr Server
 No configuration.
 
-.Solr Server System Properties (default values)
- ----
- start.solr=true
- solr.client=HttpSolrClient
- solr.http.port=8994
- ----
```
Comment (ahoffer) [inline distribution/docs/src/main/resources/content/_sources/solr-catalog-provider.adoc]: It was not a good idea to document the settings both here and in global-setting.adoc. Removed.

File: distribution/docs/src/main/resources/content/_installing/solr-service.adoc [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+. Add the argument `start` in the window pane and select *Next*.
+. Review the settings and select *Finish*.
+
+It may be necessary to update the *Security Options* under the task *Properties* to
+*Run with highest privileges* or setting user to "SYSTEM".
+
+Additionally, the process can be set to restart if it fails. The option can be found in the
+ the *Properties* > *Settings* tab.
+
+Depending on the system it may also make sense to delay the process from starting for a
```
Comment (lessarderic) [inline distribution/docs/src/main/resources/content/_installing/solr-service.adoc]: ❓Should we recommend that the delay by shorter than the one used for DDF?

File: distribution/ddf-common/src/main/resources-filtered/etc/system.properties [hunk 10]
```diff
@@ -80,20 +79,20 @@ org.codice.ddf.system.organization=${organization.name}
...
-solr.data.dir=_DO_NOT_EXPAND_${karaf.home}/solr/server/solr
+# Solr Server Provider
+solr.start=true
 solr.client=HttpSolrClient
+solr.data.dir=_DO_NOT_EXPAND_${karaf.home}/solr/server/solr
 solr.http.port=8994
 solr.http.protocol=https
-
-# Do NOT update the solr.http.url when quick installing on a remote headless server.
+solr.mem=2g
```
Comment (lessarderic) [inline distribution/ddf-common/src/main/resources-filtered/etc/system.properties]: ❓Did we lower the amount of memory allocated to the DDF Java process to account for the increased amount of memory allocated to Solr?

File: distribution/docs/src/main/resources/content/_installing/solr-service.adoc (hunk 9 above)
Comment (ahoffer) [reply distribution/docs/src/main/resources/content/_installing/solr-service.adoc]: If there was a history of problems caused by users setting Solr timeout >= DDF timeout, I would definitely add it. Without evidence, I feel adding that to our documentation is not appropriate.

File: distribution/ddf-common/src/main/resources-filtered/etc/system.properties (hunk 10 above)
Comment (ahoffer) [reply distribution/ddf-common/src/main/resources-filtered/etc/system.properties]: No, I do not believe we made any changes. 
I would not recommend making any changes until we have some operational definition of "enough memory".  For example:

* We take a particular data set and make it the "reference set".  
* Next, we define an operation(s) on the data. For instance, "ingest all data". 
* Then we define failure criteria like "Java Out of Memory exception" or "takes longer than an hour".
* Next, we experiment. We keep ratching down the memory available to DDF and Solr until the failure occurs. 
* Finally, we publish the "minimum memory requirements" as the lowest values that successfully ingested the reference data.

File: distribution/docs/src/main/resources/content/_installing/solr-service.adoc (hunk 4 above)
Comment (figliold) [reply distribution/docs/src/main/resources/content/_installing/solr-service.adoc]: I think the wording is okay, but shouldn't Unix have a capital 'U'?

File: distribution/ddf-common/src/main/resources/bin/ddf.bat [hunk 11]
```diff
@@ -7,79 +7,47 @@ PUSHD %DIRNAME%..
 SET DDF_HOME=%CD%
 POPD
 
-SET GET_PROPERTY=%DIRNAME%get_property.bat
-SET SOLR_EXEC=%DDF_HOME%\solr\bin\solr.cmd
+IF "%COMMAND%"=="" (
```
Comment (mcalcote) [inline distribution/ddf-common/src/main/resources/bin/ddf.bat]: I don't think we want to prevent the user from starting ddf with just `ddf.bat`.

Comment (mcalcote) [reply distribution/ddf-common/src/main/resources/bin/ddf.bat]: Also... we never define COMMAND, so DDF cannot be started

**Comment by ahoffer** [comment]
build now

File: distribution/docs/src/main/resources/content/_installing/file-system-access.adoc [hunk 12]
```diff
@@ -66,3 +66,58 @@ Set directory permissions to protect the ${branding} from unauthorized access.
 ** `chmod -R o-rwx ${home_directory}/etc ${home_directory}/data ${home_directory}/instances`
 ****
 
+===== Configuring Memory Allocation for the ${branding} Java Virtual Machine
+
+[[_jvm_memory_configuration]]
+The amount of memory allocated to the Java Virtual Machine host ${branding} bu the operating
```
Comment (ricklarsen) [inline distribution/docs/src/main/resources/content/_installing/file-system-access.adoc]: *by* the operating system?

File: distribution/docs/src/main/resources/content/_installing/file-system-access.adoc [hunk 13]
```diff
@@ -66,3 +66,58 @@ Set directory permissions to protect the ${branding} from unauthorized access.
...
+Remove -XX:+DisableAttachMechanism from JAVA_OPTS
+${home_directory}/bin/setenv-windows-wrapper.conf
+Comment out the -XX:+DisableAttachMechanism line and re-number remainder lines appropriately
+----
+
+===== Configuring Memory for the Solr Server
+
+[NOTE]
+====
+This section applies only if to configurations that manage the lifecycle of the Solr server.
```
Comment (ricklarsen) [inline distribution/docs/src/main/resources/content/_installing/file-system-access.adoc]: "applies only _to_ configurations?"

File: distribution/docs/src/main/resources/content/_running/ddf-service.adoc [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+==== {title}
+
+===== Running as a Service with Automatic Start on System Boot
+
+Because ${branding} is built on top of Apache Karaf, ${branding} can use the Karaf Wrapper to run ${branding} as a service and enable automatic startup and shutdown.
+When ${branding} is started using Karaf Wrapper, new `wrapper.log` and `wrapper.log.n` (where n goes from 1 to 5 by default) log files will be generated to include wrapper and console specific information.
+
+[WARNING]
+====
+When installing as a service on *NIX, do not use spaces in the path for ${home_directory} as the service scripts that are generated by the wrapper can not handle spaces.
```
Comment (ricklarsen) [inline distribution/docs/src/main/resources/content/_running/ddf-service.adoc]: cannot

File: distribution/ddf-common/src/main/resources/bin/ddf.bat [hunk 15]
```diff
@@ -8,58 +8,21 @@ SET DDF_HOME=%CD%
...
+SET SOLR_EXEC=%DDF_HOME%\bin\ddfsolr.bat
 
-:restart
+:RESTART
 REM Remove the restart file indicator so we can detect later if restart was requested
 IF EXIST "%DIRNAME%restart.jvm" DEL "%DIRNAME%restart.jvm"
 
 REM Get Solr start property
-CALL %GET_PROPERTY% start.solr true
+CALL %GET_PROPERTY% start.solr
```
Comment (mcalcote) [inline distribution/ddf-common/src/main/resources/bin/ddf.bat]: Calls to the get_propetry.bat should have a second argument that provides a default return value in the event that a property can't be found.

Comment (mcalcote) [reply distribution/ddf-common/src/main/resources/bin/ddf.bat]: This is currently `solr.start` in the `system.properties`. We can update either.

File: distribution/ddf-common/src/main/resources/bin/ddf.bat [hunk 16]
```diff
@@ -80,6 +43,6 @@ EXIT /B
 :STOP_SOLR
 IF "%SOLR_CLIENT%" == "HttpSolrClient" (
```
Comment (mcalcote) [inline distribution/ddf-common/src/main/resources/bin/ddf.bat]: Looks like this bug was already present, but this solr_client lost its definition in a previous change meaning that we never stop solr on shutdown.

File: distribution/ddf-common/src/main/resources/bin/ddf.bat (hunk 15 above)
Comment (ahoffer) [reply distribution/ddf-common/src/main/resources/bin/ddf.bat]: Or we just hardcode the value in the script and not look system.properties.

File: distribution/ddf-common/src/main/resources/bin/ddf.bat (hunk 16 above)
Comment (ahoffer) [reply distribution/ddf-common/src/main/resources/bin/ddf.bat]: Good catch!

File: distribution/ddf-common/src/main/resources/bin/ddfsolr.bat [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+    REM    the same time. They are mutually exclusive and Jetty will select one of them which
+    REM    may not be what you expect.
+    CALL SET SOLR_SSL_WANT_CLIENT_AUTH=false
+)
+
+IF "%COMMAND%"=="" ECHO Missing command. Use start, restart, stop.
+
+IF "%COMMAND%"=="start" (
+    IF "!solr.http.protocol!"=="http" ECHO **** USING INSECURE SOLR CONFIGURATION ****
+    IF "!solr.http.protocol!"=="https" ECHO Using Solr secure configuration
```
Comment (mcalcote) [inline distribution/ddf-common/src/main/resources/bin/ddfsolr.bat]: We can probably get rid of line 48, but that is a nit. 
Line 49 has 1 too many tabs (also nit)

We never seem to call `start`, do we want this block at all?

File: distribution/ddf-common/src/main/resources-filtered/etc/system.properties [hunk 18]
```diff
@@ -80,20 +79,20 @@ org.codice.ddf.system.organization=${organization.name}
...
-# Solr Client Configuration
+# Solr Provider Configuration
+# Uncomment or comment the properties based on the desired configuration:
 #
-# Uncomment the properties based on the desired configuration:
 
-# For Solr Server
-solr.data.dir=_DO_NOT_EXPAND_${karaf.home}/solr/server/solr
+# Solr Server Provider
+solr.start=true
```
Comment (rzwiefel) [inline distribution/ddf-common/src/main/resources-filtered/etc/system.properties]: I thought this made more sense up at the top along with the `start.ddf`. I thought having all the high level on/off toggles at the top (while all the specific settings down in the weeds below) would be simplest for people to use.

Comment (ahoffer) [reply distribution/ddf-common/src/main/resources-filtered/etc/system.properties]: My only defense is that I thought it made more sense with the other Solr properties. ¯\_(ツ)_/¯

File: distribution/docs/src/main/resources/content/_running/solr-service.adoc [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+Depending on the system it may also make sense to delay the process from starting for a
+few minutes until the machine has fully booted. Open the task's *Properties* settings and
+
+ .. Select *Triggers*.
+ .. Select *Edit*.
+ .. Select *Advanced Settings*.
+ .. Select *Delay Task*.
+
+
+===== Configure Solr as a Systemd Service
```
Comment (clockard) [inline distribution/docs/src/main/resources/content/_running/solr-service.adoc]: Might consider making this `Configure Solr as a *nix Systemd Service

**Pull request merged**